vm/core/context.go

```diff
@@ -72,36 +134,80 @@ func (c *Context) Handler() Handler {
 }
 
 // Spawn account.
-func (c *Context) Spawn(args scale.Encodable) error {
-	account, err := c.load(ComputePrincipal(c.Header.TemplateAddress, args))
+func (c *Context) Spawn(template Address, blob []byte) (Address, error) {
+	// calculate new principal address
+	principalAddress := ComputePrincipalFromBlob(template, blob)
+
+	// check if the account is already spawned
+	account, err := c.load(principalAddress)
 	if err != nil {
-		return err
+		return Address{}, err
 	}
+	// the account is already spawned and contains different code. this should not happen.
+	// if len(account.State) > 0 && !bytes.Equal(account.State, blob) {
+
+	// do not allow an already-spawned account to be spawned again
 	if account.TemplateAddress != nil {
-		return ErrSpawned
-	}
-	handler := c.Registry.Get(c.Header.TemplateAddress)
-	if handler == nil {
-		return fmt.Errorf("%w: spawn is called with unknown handler", ErrInternal)
+		return Address{}, ErrSpawned
 	}
-	buf := bytes.NewBuffer(nil)
-	instance, err := handler.New(args)
+
+	account.State = blob
+	account.TemplateAddress = &template
+	c.change(account)
+	c.Logger.Debug(
+		"spawn",
+		zap.String("address", principalAddress.String()),
+		zap.String("template", template.String()),
```

Comment:
Zap allows using the Stringer interface for lazy evaluation:
```suggestion
		zap.Stringer("address", principalAddress),
		zap.Stringer("template", template),
```